--- origsrc/kde-workspace-4.10.5/kwin/clients/aurorae/src/CMakeLists.txt	2013-06-28 12:10:45.067386000 -0500
+++ src/kde-workspace-4.10.5/kwin/clients/aurorae/src/CMakeLists.txt	2013-08-27 15:35:32.836260200 -0500
@@ -25,7 +25,7 @@ set(decoration_plugin_SRCS
 
 qt4_automoc(${decoration_plugin_SRCS})
 
-add_library(decorationplugin SHARED ${decoration_plugin_SRCS})
+add_library(decorationplugin MODULE ${decoration_plugin_SRCS})
 target_link_libraries(decorationplugin ${QT_QTCORE_LIBRARY} ${QT_QTDECLARATIVE_LIBRARY} ${KDE4_KDEUI_LIBRARY} kdecorations)
 install(TARGETS decorationplugin DESTINATION ${IMPORTS_INSTALL_DIR}/org/kde/kwin/decoration)
 
--- origsrc/kde-workspace-4.10.5/kwin/clients/aurorae/themes/plastik/code/CMakeLists.txt	2013-06-28 12:10:44.365416800 -0500
+++ src/kde-workspace-4.10.5/kwin/clients/aurorae/themes/plastik/code/CMakeLists.txt	2013-08-27 15:34:42.289369100 -0500
@@ -5,7 +5,7 @@ set(plastik_plugin_SRCS
 
 qt4_automoc(${plastik_plugin_SRCS})
 
-add_library(plastikplugin SHARED ${plastik_plugin_SRCS})
+add_library(plastikplugin MODULE ${plastik_plugin_SRCS})
 target_link_libraries(plastikplugin ${QT_QTCORE_LIBRARY} ${QT_QTDECLARATIVE_LIBRARY} ${KDE4_KDEUI_LIBRARY} kdecorations)
 install(TARGETS plastikplugin DESTINATION ${IMPORTS_INSTALL_DIR}/org/kde/kwin/decorations/plastik)
 install(FILES qmldir DESTINATION ${IMPORTS_INSTALL_DIR}/org/kde/kwin/decorations/plastik)
