--- origsrc/kde-workspace-4.11.18/kwin/libkwineffects/kwinglutils.cpp	2015-04-08 02:30:11.000000000 -0500
+++ src/kde-workspace-4.11.18/kwin/libkwineffects/kwinglutils.cpp	2015-05-07 10:55:06.698794300 -0500
@@ -45,6 +45,7 @@ along with this program.  If not, see <h
 #include <QVarLengthArray>
 
 #include <math.h>
+#include <strings.h>
 
 #define DEBUG_GLRENDERTARGET 0
 
--- origsrc/kde-workspace-4.10.5/kwin/scripting/scriptingutils.cpp	2013-06-28 12:10:44.442413400 -0500
+++ src/kde-workspace-4.10.5/kwin/scripting/scriptingutils.cpp	2013-08-26 21:12:49.986961700 -0500
@@ -31,6 +31,7 @@ bool validateParameters(QScriptContext *
     return true;
 }
 
+#if 0
 template<>
 bool validateArgumentType<QVariant>(QScriptContext *context, int argument)
 {
@@ -42,5 +43,6 @@ bool validateArgumentType<QVariant>(QScr
     }
     return result;
 }
+#endif
 
 }
