--- origsrc/kde-workspace-4.11.22/libs/CMakeLists.txt	2016-11-09 09:44:39.145985100 -0600
+++ src/kde-workspace-4.11.22/libs/CMakeLists.txt	2016-11-09 11:11:45.782581200 -0600
@@ -1,19 +1,19 @@
-add_subdirectory(kephal)
+#add_subdirectory(kephal)
 add_subdirectory(kworkspace)
-add_subdirectory(oxygen)
-add_subdirectory(plasmaclock)
+#add_subdirectory(oxygen)
+#add_subdirectory(plasmaclock)
 
 if(${KDE_PLATFORM_PROFILE} STREQUAL "Desktop")
-   add_subdirectory(kdm)
-   add_subdirectory(plasmagenericshell)
+#   add_subdirectory(kdm)
+#   add_subdirectory(plasmagenericshell)
    if(NOT WIN32)
-        add_subdirectory(taskmanager)
+#        add_subdirectory(taskmanager)
    endif(NOT WIN32)
 endif(${KDE_PLATFORM_PROFILE} STREQUAL "Desktop")
 
 if(NOT WIN32)
     if (NOT CMAKE_SYSTEM_NAME MATCHES Darwin)
-        add_subdirectory(ksysguard)
+#        add_subdirectory(ksysguard)
     endif (NOT CMAKE_SYSTEM_NAME MATCHES Darwin)
 endif(NOT WIN32)
 
