--- origsrc/kdebase-workspace-4.2.0/libs/CMakeLists.txt	2009-01-01 10:25:42.000000000 -0600
+++ src/kdebase-workspace-4.2.0/libs/CMakeLists.txt	2009-01-29 19:32:41.158613500 -0600
@@ -5,7 +5,9 @@ add_subdirectory(solid)
 
 if(NOT WIN32)
  add_subdirectory(taskmanager)
+if(BUILD_ksysguard)
  add_subdirectory(ksysguard)
+endif(BUILD_ksysguard)
  add_subdirectory(kworkspace)
 endif(NOT WIN32)
 
--- origsrc/kdebase-workspace-4.2.0/libs/ksysguard/processcore/processes_local_p.cpp	2008-01-04 17:55:30.000000000 -0600
+++ src/kdebase-workspace-4.2.0/libs/ksysguard/processcore/processes_local_p.cpp	2009-01-29 14:32:41.033613500 -0600
@@ -32,6 +32,12 @@
 #include <qglobal.h>
 #ifdef Q_OS_SOLARIS
 #include "processes_solaris_p.cpp"
+#else
+#include "processes_local_p.h"
+KSysGuard::ProcessesLocal::ProcessesLocal()
+{
+
+}
 #endif
 #endif
 
--- origsrc/kdebase-workspace-4.2.0/plasma/dataengines/CMakeLists.txt	2008-12-21 14:19:13.000000000 -0600
+++ src/kdebase-workspace-4.2.0/plasma/dataengines/CMakeLists.txt	2009-01-29 19:32:24.033613500 -0600
@@ -20,7 +20,9 @@ endif(KDE4_SYNDICATION_LIBS)
 
 if(NOT WIN32)
    add_subdirectory(mouse)
+if(BUILD_ksysguard)
    add_subdirectory(systemmonitor)
+endif(BUILD_ksysguard)
    add_subdirectory(tasks)
 endif(NOT WIN32)
 
