A #define in xproto breaks the KDE headers

--- origsrc/kdebase-workspace-4.3.4/kcontrol/input/main.cpp	2008-01-04 17:55:14.000000000 -0600
+++ src/kdebase-workspace-4.3.4/kcontrol/input/main.cpp	2010-01-15 03:11:06.439852500 -0600
@@ -26,14 +26,13 @@
 #include <klocale.h>
 #include <kglobal.h>
 #include <kconfig.h>
+#include <ktoolinvocation.h>
+#include <klauncher_iface.h>
 #include <QFile>
 
 #include "mouse.h"
 #include <QX11Info>
 
-#include <ktoolinvocation.h>
-#include <klauncher_iface.h>
-
 #include <X11/Xlib.h>
 #ifdef HAVE_XCURSOR
 #  include <X11/Xcursor/Xcursor.h>
