--- origsrc/kdebase-workspace-4.4.2/qguiplatformplugin_kde/CMakeLists.txt	2009-12-04 17:09:28.000000000 -0600
+++ src/kdebase-workspace-4.4.2/qguiplatformplugin_kde/CMakeLists.txt	2010-05-04 10:42:31.096382200 -0500
@@ -4,7 +4,7 @@ set(kde_SOURCES  qguiplatformplugin_kde.
 qt4_generate_moc(qguiplatformplugin_kde.cpp ${CMAKE_CURRENT_BINARY_DIR}/qguiplatformplugin_kde.moc)
 
 
-add_library (kde SHARED ${kde_SOURCES})
+add_library (kde MODULE ${kde_SOURCES})
 
 target_link_libraries(kde ${KDE4_KIO_LIBS} ${QT_QTGUI_LIBRARY})
 
