--- origsrc/kdebase-workspace-4.3.4/kmenuedit/CMakeLists.txt	2008-10-09 04:46:43.000000000 -0500
+++ src/kdebase-workspace-4.3.4/kmenuedit/CMakeLists.txt	2010-01-15 03:32:43.001011500 -0600
@@ -24,7 +24,7 @@ kde4_add_kdeinit_executable( kmenuedit $
 target_link_libraries(kdeinit_kmenuedit  ${KDE4_KDE3SUPPORT_LIBS} ${QT_QTXML_LIBRARY}
 )
 
-install(TARGETS kdeinit_kmenuedit  DESTINATION ${LIB_INSTALL_DIR} )
+install(TARGETS kdeinit_kmenuedit  DESTINATION ${INSTALL_TARGETS_DEFAULT_ARGS} )
 
 target_link_libraries( kmenuedit kdeinit_kmenuedit )
 install(TARGETS kmenuedit ${INSTALL_TARGETS_DEFAULT_ARGS})
--- origsrc/kdebase-workspace-4.4.2/plasma/desktop/shell/CMakeLists.txt
+++ src/kdebase-workspace-4.4.2/plasma/desktop/shell/CMakeLists.txt
@@ -68,7 +68,7 @@ endif(UNIX AND Q_WS_X11)
 
 target_link_libraries(kdeinit_plasma-desktop kephal)
 
-install(TARGETS kdeinit_plasma-desktop DESTINATION ${LIB_INSTALL_DIR})
+install(TARGETS kdeinit_plasma-desktop DESTINATION ${INSTALL_TARGETS_DEFAULT_ARGS})
 install(TARGETS plasma-desktop ${INSTALL_TARGETS_DEFAULT_ARGS})
 install(FILES plasma.desktop plasma-desktop.desktop DESTINATION ${AUTOSTART_INSTALL_DIR})
 install(FILES kcfg/plasma-shell-desktop.kcfg  DESTINATION  ${KCFG_INSTALL_DIR})
--- origsrc/kdebase-workspace-4.5.5/plasma/generic/shells/plasma-windowed/CMakeLists.txt	2010-05-16 05:04:32.000000000 -0500
+++ src/kdebase-workspace-4.5.5/plasma/generic/shells/plasma-windowed/CMakeLists.txt	2011-01-14 02:17:00.780058600 -0600
@@ -14,6 +14,6 @@ kde4_add_kdeinit_executable(plasma-windo
 target_link_libraries(kdeinit_plasma-windowed ${KDE4_PLASMA_LIBS} kworkspace ${KDE4_KIO_LIBS} ${X11_LIBRARIES} ${KDE4_KFILE_LIBS} )
 
 
-install(TARGETS kdeinit_plasma-windowed DESTINATION ${LIB_INSTALL_DIR})
+install(TARGETS kdeinit_plasma-windowed ${INSTALL_TARGETS_DEFAULT_ARGS})
 install(TARGETS plasma-windowed ${INSTALL_TARGETS_DEFAULT_ARGS})
 
--- origsrc/kdebase-workspace-4.4.2/plasma/netbook/shell/CMakeLists.txt	2009-10-26 04:14:50.000000000 -0500
+++ src/kdebase-workspace-4.4.2/plasma/netbook/shell/CMakeLists.txt	2010-05-04 12:19:35.754533600 -0500
@@ -19,7 +19,7 @@ if(X11_Xrender_FOUND)
    target_link_libraries(kdeinit_plasma-netbook ${X11_Xrender_LIB})
 endif(X11_Xrender_FOUND)
 
-install(TARGETS kdeinit_plasma-netbook DESTINATION ${LIB_INSTALL_DIR})
+install(TARGETS kdeinit_plasma-netbook ${INSTALL_TARGETS_DEFAULT_ARGS})
 install(TARGETS plasma-netbook ${INSTALL_TARGETS_DEFAULT_ARGS})
 install(FILES plasma-default-layoutrc DESTINATION ${DATA_INSTALL_DIR}/plasma-netbook/)
 #install(FILES plasma-netbook.desktop DESTINATION ${AUTOSTART_INSTALL_DIR})
--- origsrc/kdebase-workspace-4.4.2/qguiplatformplugin_kde/CMakeLists.txt	2009-12-04 17:09:28.000000000 -0600
+++ src/kdebase-workspace-4.4.2/qguiplatformplugin_kde/CMakeLists.txt	2010-05-04 10:42:31.096382200 -0500
@@ -4,7 +4,7 @@ set(kde_SOURCES  qguiplatformplugin_kde.
 qt4_generate_moc(qguiplatformplugin_kde.cpp ${CMAKE_CURRENT_BINARY_DIR}/qguiplatformplugin_kde.moc)
 
 
-add_library (kde SHARED ${kde_SOURCES})
+add_library (kde MODULE ${kde_SOURCES})
 
 target_link_libraries(kde ${KDE4_KIO_LIBS} ${QT_QTGUI_LIBRARY})
 
